accepted/2022/simplify-output-paths.md

```diff
@@ -0,0 +1,58 @@
+# Simplify .NET Output Paths
+
+We'd like to simplify the output paths for .NET 8.
+
+Currently, the default output path for a .NET project includes folders for the `TargetFramework` and the `RuntimeIdentifier` (if it is set).  Additionally, the publish output goes in a `publish` subfolder of the output folder.  The resulting paths are as follows:
+
+- `bin\<Configuration>\<TargetFramework>` - Build output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>` - Build output with `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\publish` - Publish output with no `RuntimeIdentifier`
+- `bin\<Configuration>\<TargetFramework>\<RuntimeIdentifier>\publish` - Publish output with `RuntimeIdentifier`
+
+This is rather complicated an inconsistent, which can make for a poor first impression of the .NET platform.
+
+We'd like to try to improve the output path structure.  Desired qualities include:
+
+- Simple and consistent
+- Avoid cases where one output path is nested inside of another one
+- Avoid excessively deep directory heirarchies
+- Get rid of `obj` folder in project root
+
+## Proposed behavior
+
+Projects targeting .NET and higher will by default use a new output path format.  The output path will consist of the following 3 nested folders
+
+- `bin` - All output (including intermediate output, will go under this folder)
+- Output Type - Such as `build`, `publish`, `obj`, or `packages`
+- Pivots - This will at minimum be the `Configuration`, such as `Debug` or `Release`. Other pivots such as `TargetFramework` or `RuntimeIdentifier` may also be included, and the pivots will be joined by the underscore (`_`) character
+  - `TargetFramework` will be included in the folder name if the project is multi-targeted (`TargetFrameworks` is non-empty), or if the `TargetFramework` property was set via the command line (ie is a global property)
+  - `RuntimeIdentifier` will be included in the folder name if it was explicitly set (either in a project file or on the command line).  If it is set automatically by the SDK (for exmaple because `SelfContained` was set)
+
+Some examples:
+
+- `bin\build\Debug` - The build output path for a simple project when you run `dotnet build`
+- `bin\obj\Debug` - The intermediate output path for a simple project when you run `dotnet build`
+- `bin\build\Debug_net8.0` - The build output path for the `net8.0` build of a multi-targeted project
+- `bin\publish\Release_linux-x64` - The publish path for a simple app when publishing for `linux-x64`
+- `bin\package\Release` - The folder where the release .nupkg will be created for a project
```

Comment:
I like the name `bin` for build binaries. Did you consider naming the root output folder ´artifacts`, similar to what we have in arcadified repositories today and keep the `bin` name for the final binaries?

Reply:
<img width="227" alt="image" src="https://user-images.githubusercontent.com/7412651/203063265-d9a5012c-592d-413d-890e-4462394beda3.png">

Reply:
It'd be good to know what other languages are doing, e.g. is there some common output path naming that we could move to?

Reply:
The issue with using a new name like `artifacts` as the first level folder is that it is a new name, so it would conflict with any projects that already a folder with that name for other purposes (such as source code).

Still might be worth considering.

Reply:
Yeah I think it is worth considering. As a small datapoint, the dotnet project structure gist by @davidfowl https://gist.github.com/davidfowl/ed7564297c61fe9ab814 also uses `artifacts` and it has 1.2k stars :)

Reply:
One thing that we might want to have in mind is that some people run into MAX_PATH issues on Windows, and making the path significantly longer could make this problem worse.

Reply:
While "artifacts" increases the length by 4 chars (bin + build = 8 chars, artifacts + bin = 12 chars), the overall proposal already shortens the length of the output/intermediate path so that shouldn't be a concern.

Reply:
for what it's worth, [the Visual Studio default `.gitignore` already ignores any `artifacts` directory](https://github.com/github/gitignore/blob/e5323759e387ba347a9d50f8b0ddd16502eb71d4/VisualStudio.gitignore#L63), and has done so [since 2015](https://github.com/github/gitignore/commit/467e257fdb71b6b920dde08ef16337e80ef91bb6). Though not everyone uses git, and not all git users use the default `.gitignore`.

Reply:
I think we're still longer paths here in some cases, like on android today we have:
`obj/Debug/net7.0-android/android-arm64/`

However if I'm reading the spec correctly, the new path would be:
`artifacts/intermediates/debug_net7.0-android_android-arm64/` 
(Yes in these cases, RID is often required since we often build mobile apps with multiple RIDs, even as part of Debug builds, for reasons).

This adds 20 more characters to the path in this case.  As mentioned, we have historically had many problems with MAX_PATH issues on windows and have taken a bunch of steps along the way to shorten them where possible, often at the expense of making it easy to understand the path structure.

RootArtifactsPath may help with path lengths (and honestly may be desirable in a lot of cases - for example it might be nice to have one `artifacts/*` path alongside the .sln that can be deleted to clear out everything for all the projects, but given that a .sln is not _required_ for building, I'm not sure we can rely on this to help with MAX_PATH issues.

Reply:
Also, some platforms have long version numbers

```
<proj_root>\obj\Debug\net7.0-windows10.0.19041.0\win10-arm64\
```

Becomes:

```
artifacts/intermediates/debug_net7.0-windows10.0.19041.0_win10-arm64/
```

Reply:
@dsplaisted ^ ping to the concerns with increasing path length

Reply:
I feel like the folder should be prefixed with a period to make sure it doesn't get sorted in the middle of project folders. Something that has always driven me a little crazy is how the current `bin` and `obj` folders don't have a `.` prefix to make them sort above the project namespace directories.

Other build frameworks I have used have a single "cache" directory which in an automated build pipeline you would set as the persistent cache directory. Personally, I think the following structure would make the most sense:
```markdown
.msbuild/
├── build/                     # This would be the "cacheable" directory
│   ├── bin/
│   │   ├── Project.Lib
│   │   ├── Project.Web
│   │   └── ...
│   └── obj/
│       ├── Project.Lib
│       ├── Project.Web
│       └── ...
├── pack/
│   ├── Project.Lib.nupkg
│   ├── Project.Lib.snupkg
│   └── ...
└── publish/
    ├── Project.Web
    └── ...
```

Reply:
With respect to the long path limitations, I've updated the proposal so that you have to opt in to the new format, and called out long path issues as one of the obstacles to enabling the new format by default.